pennylane/_qubit_device.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QubitDevice` abstract base class.
+"""
+
+# For now, arguments may be different from the signatures provided in Device
+# e.g. instead of expval(self, observable, wires, par) have expval(self, observable)
+# pylint: disable=arguments-differ, abstract-method, no-value-for-parameter
+
+import itertools
+from collections import OrderedDict
+import numpy as np
+
+from pennylane.operation import Sample, Variance, Expectation, Probability
+from pennylane.qnodes import QuantumFunctionError
+from pennylane import Device
+
+
+class QubitDevice(Device):
+    """Abstract base class for PennyLane qubit devices.
+
+    Args:
+        wires (int): number of subsystems in the quantum state represented by the device.
+            Default 1 if not specified.
+        shots (int): Number of circuit evaluations/random samples used to estimate
+            expectation values of observables. Defaults to 1000 if not specified.
+    """
+    #pylint: disable=too-many-public-methods
+    def __init__(self, wires=1, shots=1000, analytic=True):
+        super().__init__(wires=wires, shots=shots)
+        self.analytic = analytic
+
+        self._state = None
+        self._memory = None
+        self._samples = None
+
+    @property
+    def _probability(self):
+        """Return the current probability of each computational basis state based on the state vector."""
+
+        return np.abs(self._state)**2
```

Comment:
It seems this is only ever called once (on line 191). Might as well just delete this for now, and explicitly have
```python
prob = self.marginal_prob(np.abs(self._state)**2, wires)
```
on line 191.